### PR TITLE
Fix broker error handling

### LINKED ISSUE #389: User cancel in Azure Authenticator app returned to client as AD_FAILED instead of AD_USER_CANCELLED
If I cancel the credentials prompt in Azure Authenticator app, I get AD_FAILED not AD_USER_CANCELLED:
2015-09-18 14:39:03.821 Word[482:127664] ADALiOS [2015-09-18 21:39:03 - BBBCCAC6-D5BA-4D3D-B493-8974935DA0A3] ERROR: Error raised: 0. Additional Information: Domain: ADBrokerResponseErrorDomain ProtocolCode:(null) Details:The user has cancelled the authorization.. ErrorCode: 0.
[ Word ] msoLogging.Liblet EnterpriseDataProtection actions: { "tag": "ibsgp", "message": "IntuneSDKLog {"IntuneSDKLog": "IntuneMAM: ADALiOS [2015-09-18 21:39:03 - BBBCCAC6-D5BA-4D3D-B493-8974935DA0A3] Error raised: 0, Domain: ADBrokerResponseErrorDomain ProtocolCode:(null) Details:The user has cancelled the authorization., error code: 0"}" }
2015-09-18 14:39:03.842 Word[482:127664] ADALiOS [2015-09-18 21:39:03 - BBBCCAC6-D5BA-4D3D-B493-8974935DA0A3] INFORMATION: (error=Error with code: 0 Domain: ADBrokerResponseErrorDomain ProtocolCode:(null) Details:The user has cancelled the authorization.. Inner error details: Error Domain=ADBrokerResponseErrorDomain Code=0 "The operation couldn’t be completed. (ADBrokerResponseErrorDomain error 0.)", mrrt=NO, status=AD_FAILED, item=(null)). Additional Information: (null). ErrorCode: 0.

ADAuthenticationResult-> resultFromBrokerResponse: treats all failures responses received from the broker app as an error, by calling “resultFromError:” when instead it should check for cancellation error code from the broker response and call “resultFromCancellation”. But the response from the broker actually doesn’t contain an error code to check for:
{
    code = "(null)";
    "correlation_id" = "6EBD8A1E-573C-45E5-9EE9-7716B068EECF";
    "error_description" = "The user has cancelled the authorization.";
}

### PR DESCRIPTION
**Pull request by RPangrle**
Combined with an update to Authenticator this should fix #389

### CODE REVIEW

File: ADALiOS/ADALiOS/ADAuthenticationResult+Internal.m [hunk 1]
```diff
@@ -111,43 +111,87 @@ +(ADAuthenticationResult*) resultFromCancellation: (NSUUID*) correlationId
...
+                                       userInfo:nil];
+    ADAuthenticationError* error = [ADAuthenticationError errorFromNSError:nsError
+                                                              errorDetails: @"No broker response received."];
+    return [ADAuthenticationResult resultFromError:error correlationId:nil];
+}
+
++ (ADAuthenticationResult*)resultForBrokerErrorResponse:(NSDictionary*)response
+{
+    NSString* correlationIdStr = [response valueForKey:OAUTH2_CORRELATION_ID_RESPONSE];
+    NSUUID* correlationId =  correlationIdStr ? [[NSUUID alloc] initWithUUIDString:correlationIdStr] : nil;
```
Comment (weijjia) [inline ADALiOS/ADALiOS/ADAuthenticationResult+Internal.m]: Can we log something saying broker doesn't return us back the correlation id?

File: ADALiOS/ADALiOS/ADAuthenticationResult+Internal.m [hunk 2]
```diff
@@ -111,43 +111,87 @@ +(ADAuthenticationResult*) resultFromCancellation: (NSUUID*) correlationId
...
+       
+        error = [ADAuthenticationError errorFromAuthenticationError:errorCode
+                                                       protocolCode:protocolCode
+                                                       errorDetails:errorDetails];
+    }
+    else
+    {
+        NSError* nsError = [NSError errorWithDomain:ADBrokerResponseErrorDomain
+                                               code:errorCode
+                                           userInfo:nil];
```
Comment (weijjia) [inline ADALiOS/ADALiOS/ADAuthenticationResult+Internal.m]: nitpick: indent

**Comment by weijjia** [comment]
looks good to me! I like those tests!! :shipit:

File: ADALiOS/ADALiOS/ADAuthenticationResult+Internal.m (hunk 2 above)
Comment (RPangrle) [inline ADALiOS/ADALiOS/ADAuthenticationResult+Internal.m]: This is actually the indent style you're supposed to use on Objective-C, the colons ( : ) line up, unless you're seeing something I don't.

**Pull request merged**